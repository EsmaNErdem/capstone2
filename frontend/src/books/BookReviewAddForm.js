import React, { useState } from 'react';
import ListItem from '@mui/material/ListItem';
import TextField from '@mui/material/TextField';
import Button from '@mui/material/Button';

/**
 * Displays input text area
 * 
 * - Send user review data to database. If fails, shows error on bookcard to user
 * 
 * - BookReviewDrawer ==> BookReviewAddForm
 */
const BookReviewAddForm = ({addReviews, rowCount=3, close=false, closeModal}) => {
    console.debug("BookReviewAddForm");

    const [formData, setFormData] = useState({ review: '' });

    // Send review to database
    const handleAddReview = async e => {
        e.preventDefault();
        await addReviews(formData);
        setFormData({ review: '' })
        if (close) closeModal();
    }

    const handleReviewChange = e => {
        setFormData(data => ({
            ...data,
            review: e.target.value || undefined
        }))
    }
    return (
        <>
            <ListItem disablePadding sx={{ marginBottom: 2 }}>
                <TextField
                    label="Add a Review"
                    multiline
                    rows={rowCount}
                    variant="outlined"
                    fullWidth
                    sx={{ width: '100%' }}
                    value={formData.review}
                    onChange={handleReviewChange}
                />
            </ListItem>
            <ListItem disablePadding>
                <Button variant="contained" onClick={handleAddReview}>
                    Add Review
                </Button>
            </ListItem>
        </>
    )
}

export default BookReviewAddForm;