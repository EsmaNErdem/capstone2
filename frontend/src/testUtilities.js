import React from "react";
import UserContext from "./auth/UserContext";

const demoUser = {
  username: "testuser",
  firstNsme: "testfirst",
  lastName: "testlast",
  email: "test@test.net",
  img: null,
};

const UserProvider =
    ({ children, currentUser = demoUser, hasLikedReview = () => false, hasLikedBook = () => false, isUserReview = () => false, hasFollowing = () => false, setCurrentUser = u => u }) => (
    <UserContext.Provider value={{ currentUser, hasLikedReview, hasLikedBook, hasFollowing, setCurrentUser, isUserReview }}>
      {children}
    </UserContext.Provider>
);

const CurrUserProvider =
    ({ children, currentUser = demoUser, hasLikedReview = () => false, hasLikedBook = () => false, isUserReview = () => true, hasFollowing = () => false, setCurrentUser = u => u }) => (
    <UserContext.Provider value={{ currentUser, hasLikedReview, hasLikedBook, hasFollowing, setCurrentUser, isUserReview }}>
      {children}
    </UserContext.Provider>
);

const NonUserProvider =
    ({ children, currentUser = null, hasLikedReview = () => false, hasLikedBook = () => false, hasFollowing = () => false }) => (
    <UserContext.Provider value={{ currentUser, hasLikedReview, hasLikedBook, hasFollowing,}}>
      {children}
    </UserContext.Provider>
);

export { UserProvider, CurrUserProvider, NonUserProvider };
