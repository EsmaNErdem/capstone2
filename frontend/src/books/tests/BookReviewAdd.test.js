import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { UserProvider } from "../../testUtilities";
import { render, fireEvent, act, waitFor } from '@testing-library/react';
import BookReviewAddForm from '../BookReviewAddForm';

const addReviews = jest.fn();
jest.useFakeTimers();

afterEach(() => {
  jest.clearAllTimers();
})

test('renders without crashing', async () => {  
    await act(async () => {
      render(
        <MemoryRouter>
          <UserProvider>
            <BookReviewAddForm addReviews={addReviews} closeModal={()=>{}} />
          </UserProvider>
        </MemoryRouter>
      );
    });
});

test('it renders and matches with snapshot', async () => {  
    await act(async () => {
      ({ asFragment } = render(
        <MemoryRouter>
          <UserProvider>
          <BookReviewAddForm addReviews={addReviews} closeModal={()=>{}} />
          </UserProvider>
        </MemoryRouter>
      ));
    });

    expect(asFragment()).toMatchSnapshot();
});

test('submit review input', async () => {  
    await act(async () => {
      ({ getByRole, getByLabelText } = render(
        <MemoryRouter>
          <UserProvider>
          <BookReviewAddForm addReviews={addReviews} closeModal={()=>{}} />
          </UserProvider>
        </MemoryRouter>
      ));
    });

    const reviewInput = getByLabelText('Add a Review');
    fireEvent.change(reviewInput, { target: { value: 'Review' } });
    const addButton = getByRole('button', { name: 'Add Review' });
    fireEvent.click(addButton);

    expect(addReviews).toHaveBeenCalledWith({ review: 'Review' });
});