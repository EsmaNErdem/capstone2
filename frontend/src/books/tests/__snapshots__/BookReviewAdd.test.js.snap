// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders and matches with snapshot 1`] = `
<DocumentFragment>
  <li
    class="MuiListItem-root MuiListItem-gutters css-1b0nbcv-MuiListItem-root"
  >
    <div
      class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-g0i1uo-MuiFormControl-root-MuiTextField-root"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="false"
        for=":r1:"
        id=":r1:-label"
      >
        Add a Review
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-multiline css-8ewcdo-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <textarea
          aria-invalid="false"
          class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline css-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
          id=":r1:"
          rows="3"
          style="height: 0px; overflow: hidden;"
        />
        <textarea
          aria-hidden="true"
          class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline css-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
          readonly=""
          style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); padding-top: 0px; padding-bottom: 0px; width: 100%;"
          tabindex="-1"
        />
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-yjsfm1"
          >
            <span>
              Add a Review
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
  </li>
  <li
    class="MuiListItem-root MuiListItem-gutters css-1nxmd3h-MuiListItem-root"
  >
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-sghohy-MuiButtonBase-root-MuiButton-root"
      tabindex="0"
      type="button"
    >
      Add Review
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </li>
</DocumentFragment>
`;
